src/main/java/com/gorlah/kappabot/function/BotResponseFormatter.java

```diff
@@ -0,0 +1,50 @@
+package com.gorlah.kappabot.function;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.text.StringSubstitutor;
+import org.springframework.stereotype.Component;
+
+import java.beans.IntrospectionException;
+import java.beans.Introspector;
+import java.beans.PropertyDescriptor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Component
+public class BotResponseFormatter {
```

Comment:
@ploober do you have any thoughts on what I'm doing here?  What would you improve about this or would you even recommend doing it this way at all?

Reply:
I don't think I would use reflection this way personally. Maybe it would be more useful if there were a ton of fields that were intended for property replacement. We have like 20 lines of code here for substituting 2 fields, which could be done without reflection by constructing the map in one line:
`Map.of("author", metadata.getAuthor(), "authorMention", metadata.getAuthorMention());`

If you really do want it this way, I would probably make the getPropertyDescriptorsFromMetadata() be something that is called once and cached or just initialized in a static field on startup so it doesn't even need to repeat the same reflection of SubstitutableMetadataFields (reflection = slower than most stuff). You could also make this a lot less fragile by replacing the cast to String on line 43 with a null check and call to .toString() since that would make it work for a much broader variety of return types like Long, Integer, etc that could be returned by methods in SubstitutableMetadataField (which also needs to be documented since it will only work as expected if the methods are getters which return String and a few other types).

Reply:
These are all fair points. My main reason for doing this were because I really like how the map is just generated automatically and adding new replaceable properties can be done without ever having to modify this class. I agree that it is probably a bit overkill when in reality it's just the one property that is being replaced, however I have been writing these changes with the intention that it would be easy to implement other properties in the future. For now I can get rid of the reflection and if it turns out we are adding a bunch more properties then maybe we can reimplement something like this in the future with the modifications you suggested.